### PR TITLE
NMC/Gr bug fix & Switch to cycle_counts vector

### PR DESCRIPTION
**Pull request by dguittet**
There was a bug in the calculation of the b1 coefficient in #748 that is fixed here. There were also some adaptations to get the Ua and Voc calculations to be more aligned. Many results in tests have changed.

This PR also changes the NMC/Gr life model to use the `cycle_counts` structure in the cycle model, rather than maintain its own `cycle_DOD_range` vector. However, the usage is slightly different from the cycle-calendar model. The latter is tracking a frequency table of cycles at the DODs given by the input cycle matrix for the whole simulation duration. Whereas the NMC/Gr model will be tracking each cycle separately (essentially the count is always 1 and the DODs are ordered by when the cycle is completed) and for one day at a time.

Needing to easily append new rows to the cycle_count matrix led to my switching that structure from a matrix_t to a 2d vector, so I also added some helper functions for translating between the two.

### CODE REVIEW

File: shared/lib_battery_lifetime_calendar_cycle.h
Comment (brtietz) [inline shared/lib_battery_lifetime_calendar_cycle.h:73]: This comment is no longer strictly true, since the NMC and LMO/LTO variables don't have a cycling matrix. Can you add some text on how the data here varies by lifetime algorithm?

File: shared/lib_util.h
Comment (brtietz) [inline shared/lib_util.h:814]: This is super helpful, thank you!

File: shared/lib_battery_lifetime_calendar_cycle.h
Comment (dguittet) [reply shared/lib_battery_lifetime_calendar_cycle.h]: Yes, definitely. Updated the comment and clarified also in the SSC table for cmod_battery_stateful.cpp

File: test/shared_test/lib_battery_lifetime_nmc_test.cpp
Comment (brtietz) [inline test/shared_test/lib_battery_lifetime_nmc_test.cpp:703]: Need to remove local paths from the tests.

**Review by brtietz — CHANGES_REQUESTED** [review]
This mechanism for cycle counting makes more sense to me, and actually clarifies some of the NMC code.

The major required change before merging is to re-work lib_battery_lifetime_nmc_test, BTMSTest to not have local paths. More comments are nice as you have time - I think the one in the calendar cycle header file will save head scratching later.

File: test/shared_test/lib_battery_lifetime_nmc_test.cpp
Comment (dguittet) [reply test/shared_test/lib_battery_lifetime_nmc_test.cpp:703]: Oops this was erroneously committed. Let me remove this

File: shared/lib_util.h
Comment (dguittet) [reply shared/lib_util.h:814]: Great!

**Comment by brtietz** [comment]
> I'm happy to add other comments where you think they would be helpful, any suggestions? I removed the BTMS test that was based on a bunch of large local files, which I had committed by accident.

The other one that came to mind was indicating what the DOD indices are in the NMC and LMO/LTO cpp files the first time they are used ex: `DOD[0] // cycle count`. I agree that this is more useful than what I wrote, but there are advantages to enums sometimes. ;)

**Review by brtietz — APPROVED** [review]
Approved, conditional on the tests passing.

**Pull request merged**
